src/main/java/gov/nasa/jpf/constraints/solvers/ConstraintSolverService.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2015 United States Government, as represented by the Administrator
+ *                of the National Aeronautics and Space Administration. All Rights Reserved.
+ *           2017-2021 The jConstraints Authors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gov.nasa.jpf.constraints.solvers;
+
+import gov.nasa.jpf.constraints.api.ConstraintSolver;
+import gov.nasa.jpf.constraints.solvers.spi.SolverProvider;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Properties;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class ConstraintSolverService {
+
+  private static ConstraintSolverService service;
+  private ServiceLoader<SolverProvider> loader;
+  private final Properties config = new Properties();
+  private final Map<String, SolverProvider> providers = new HashMap<String, SolverProvider>();
+  protected static Logger logger = Logger.getLogger("constraints");
+
+  private ConstraintSolverService() {
+    loader = ServiceLoader.load(SolverProvider.class);
+    discoverProviders();
+  }
+
+  public static synchronized ConstraintSolverService getInstance() {
+    if (service == null) {
+      service = new ConstraintSolverService();
+    }
+    return service;
+  }
+
+  public String[] getNames() {
+    String[] stringArray = {};
+    return providers.keySet().toArray(stringArray);
+  }
```

Comment:
Arrays are fairly unwieldy to use, why not return a `Collection<String>` or a `Set<String>`?

```suggestion
  public Collection<String> getNames() {
    return new unmodifiableSet(HashSet(providers.keySet())); // or similar
  }
```